File: frontend/src/types/prescription.ts
/**
 * Prescription Types and Interfaces
 *
 * TypeScript type definitions for prescription-related data structures,
 * matching the backend Rust models for medication prescriptions in the
 * medical practice management system.
 *
 * All medication data is encrypted at rest using AES-256-GCM.
 */

/**
 * Prescription status enum
 */
export enum PrescriptionStatus {
  /** Prescription is active */
  ACTIVE = 'ACTIVE',
  /** Prescription has been completed (all refills used) */
  COMPLETED = 'COMPLETED',
  /** Prescription was cancelled */
  CANCELLED = 'CANCELLED',
  /** Prescription has been discontinued by provider */
  DISCONTINUED = 'DISCONTINUED',
  /** Prescription is temporarily on hold */
  ON_HOLD = 'ON_HOLD',
}

/**
 * Medication form enum
 */
export enum MedicationForm {
  TABLET = 'TABLET',
  CAPSULE = 'CAPSULE',
  LIQUID = 'LIQUID',
  SYRUP = 'SYRUP',
  SUSPENSION = 'SUSPENSION',
  INJECTION = 'INJECTION',
  TOPICAL = 'TOPICAL',
  CREAM = 'CREAM',
  OINTMENT = 'OINTMENT',
  GEL = 'GEL',
  PATCH = 'PATCH',
  INHALER = 'INHALER',
  DROPS = 'DROPS',
  SUPPOSITORY = 'SUPPOSITORY',
  OTHER = 'OTHER',
}

/**
 * Route of administration enum
 */
export enum RouteOfAdministration {
  ORAL = 'ORAL',
  TOPICAL = 'TOPICAL',
  INTRAVENOUS = 'INTRAVENOUS',
  INTRAMUSCULAR = 'INTRAMUSCULAR',
  SUBCUTANEOUS = 'SUBCUTANEOUS',
  SUBLINGUAL = 'SUBLINGUAL',
  RECTAL = 'RECTAL',
  INHALATION = 'INHALATION',
  OPHTHALMIC = 'OPHTHALMIC',
  OTIC = 'OTIC',
  NASAL = 'NASAL',
  TRANSDERMAL = 'TRANSDERMAL',
  OTHER = 'OTHER',
}

/**
 * Drug interaction warning severity
 * Extended to include contraindicated and unknown from DDInter database
 */
export type DrugInteractionSeverity = 'contraindicated' | 'major' | 'moderate' | 'minor' | 'unknown';

/**
 * Drug interaction warning structure (simple format for display)
 */
export interface DrugInteractionWarning {
  medication_name: string;
  severity: DrugInteractionSeverity;
  description: string;
}

/**
 * Full drug interaction record from backend API
 */
export interface DrugInteraction {
  id: string;
  drug_a_atc_code: string;
  drug_a_name?: string;
  drug_b_atc_code: string;
  drug_b_name?: string;
  severity: DrugInteractionSeverity;
  effect?: string;
  mechanism?: string;
  management?: string;
  source: string;
}

/**
 * Request to check interactions between multiple medications
 */
export interface CheckInteractionsRequest {
  atc_codes: string[];
  min_severity?: string;
  include_inactive?: boolean;
}

/**
 * Request to check interactions when adding a new medication (by ATC code)
 */
export interface CheckNewMedicationRequest {
  new_atc_code: string;
  existing_atc_codes: string[];
  min_severity?: string;
}

/**
 * Request to check interactions when adding a new medication for a patient
 * Uses medication name with fuzzy matching instead of ATC codes
 */
export interface CheckNewMedicationForPatientRequest {
  new_medication_name: string;
  new_generic_name?: string;
  patient_id: string;
  min_severity?: string;
}

/**
 * Response from drug interaction check endpoints
 */
export interface CheckInteractionsResponse {
  interactions: DrugInteraction[];
  total: number;
  major_count: number;
  moderate_count: number;
  minor_count: number;
  highest_severity?: DrugInteractionSeverity;
}

/**
 * Statistics about the drug interaction database
 */
export interface InteractionStatistics {
  total: number;
  contraindicated: number;
  major: number;
  moderate: number;
  minor: number;
  unknown: number;
  sources: number;
}

/**
 * Main Prescription interface (DTO from backend)
 */
export interface Prescription {
  id: string;

  // References
  visit_id?: string;
  patient_id: string;
  provider_id: string;

  // Medication details (encrypted)
  medication_name: string;
  generic_name?: string;
  dosage: string;
  form?: MedicationForm;
  route?: RouteOfAdministration;
  frequency: string;
  duration?: string;
  quantity?: number;
  refills: number;

  // Instructions (encrypted)
  instructions?: string;
  pharmacy_notes?: string;

  // Status
  status: PrescriptionStatus;
  prescribed_date: string; // ISO 8601 format
  start_date?: string;
  end_date?: string;
  discontinuation_reason?: string;
  discontinued_at?: string;
  discontinued_by?: string;

  // Drug interactions (stored as JSONB)
  interaction_warnings?: DrugInteractionWarning[];

  // Audit
  created_at: string;
  updated_at: string;
}

/**
 * Request to create a new prescription
 */
export interface CreatePrescriptionRequest {
  visit_id?: string;
  patient_id: string;
  provider_id: string;
  medication_name: string;
  generic_name?: string;
  dosage: string;
  form?: MedicationForm;
  route?: RouteOfAdministration;
  frequency: string;
  duration?: string;
  quantity?: number;
  refills?: number;
  instructions?: string;
  pharmacy_notes?: string;
  prescribed_date: string; // ISO 8601 format
  start_date?: string;
  end_date?: string;
  /** Drug interaction warnings detected at creation time (stored for display in lists) */
  interaction_warnings?: DrugInteractionWarning[];
}

/**
 * Request to update an existing prescription
 */
export interface UpdatePrescriptionRequest {
  dosage?: string;
  frequency?: string;
  duration?: string;
  quantity?: number;
  refills?: number;
  instructions?: string;
  pharmacy_notes?: string;
  status?: PrescriptionStatus;
  start_date?: string;
  end_date?: string;
}

/**
 * Request to discontinue a prescription
 */
export interface DiscontinuePrescriptionRequest {
  reason: string;
}

/**
 * Medication search result
 */
export interface MedicationSearchResult {
  name: string;
  generic_name?: string;
  common_dosages?: string[];
  common_forms?: MedicationForm[];
  default_route?: RouteOfAdministration;
}

/**
 * Request to create a custom medication
 */
export interface CreateCustomMedicationRequest {
  name: string;
  generic_name?: string;
  form?: string;
  dosage_strength?: string;
  route?: string;
  common_dosages?: string[];
  notes?: string;
}

/**
 * Response from creating a custom medication
 */
export interface CreateCustomMedicationResponse {
  id: string;
  message: string;
}

/**
 * Search/filter parameters for prescriptions
 */
export interface PrescriptionSearchFilters {
  patient_id?: string;
  provider_id?: string;
  status?: PrescriptionStatus;
  medication_name?: string;
  start_date?: string; // ISO 8601 format
  end_date?: string; // ISO 8601 format
  limit?: number;
  offset?: number;
}

/**
 * Paginated prescription list response
 */
export interface PrescriptionListResponse {
  prescriptions: Prescription[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * Prescription template for quick prescribing
 */
export interface PrescriptionTemplate {
  id: string;
  name: string;
  description?: string;
  medication_name: string;
  generic_name?: string;
  dosage: string;
  form?: MedicationForm;
  route?: RouteOfAdministration;
  frequency: string;
  duration?: string;
  quantity?: number;
  refills: number;
  instructions?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * Request to create a prescription template
 */
export interface CreatePrescriptionTemplateRequest {
  name: string;
  description?: string;
  medication_name: string;
  generic_name?: string;
  dosage: string;
  form?: MedicationForm;
  route?: RouteOfAdministration;
  frequency: string;
  duration?: string;
  quantity?: number;
  refills?: number;
  instructions?: string;
}

/**
 * Helper Functions
 */

/**
 * Check if prescription can be refilled
 */
export function canRefill(status: PrescriptionStatus): boolean {
  return status === PrescriptionStatus.ACTIVE;
}

/**
 * Check if prescription can be discontinued
 */
export function canDiscontinue(status: PrescriptionStatus): boolean {
  return [PrescriptionStatus.ACTIVE, PrescriptionStatus.ON_HOLD].includes(status);
}

/**
 * Get status display color for UI
 */
export function getStatusColor(status: PrescriptionStatus): string {
  switch (status) {
    case PrescriptionStatus.ACTIVE:
      return 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-200';
    case PrescriptionStatus.COMPLETED:
      return 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-200';
    case PrescriptionStatus.CANCELLED:
      return 'bg-red-100 text-red-800 dark:bg-red-900 dark:text-red-200';
    case PrescriptionStatus.DISCONTINUED:
      return 'bg-orange-100 text-orange-800 dark:bg-orange-900 dark:text-orange-200';
    case PrescriptionStatus.ON_HOLD:
      return 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900 dark:text-yellow-200';
    default:
      return 'bg-gray-100 text-gray-800';
  }
}

/**
 * Get interaction severity display color for UI (badge style with solid background)
 * Colors are consistent with DrugInteractionWarning component
 */
export function getInteractionSeverityColor(severity: DrugInteractionSeverity): string {
  switch (severity) {
    case 'contraindicated':
      return 'bg-purple-600 text-white dark:bg-purple-700';
    case 'major':
      return 'bg-red-600 text-white dark:bg-red-700';
    case 'moderate':
      return 'bg-amber-500 text-white dark:bg-amber-600';
    case 'minor':
      return 'bg-blue-500 text-white dark:bg-blue-600';
    case 'unknown':
    default:
      return 'bg-gray-500 text-white dark:bg-gray-600';
  }
}

/**
 * Format prescription for display (e.g., "Aspirin 100mg, 1 tablet daily")
 */
export function formatPrescription(prescription: Prescription): string {
  const parts: string[] = [prescription.medication_name];

  if (prescription.dosage) {
    parts.push(prescription.dosage);
  }

  if (prescription.form) {
    parts.push(prescription.form.toLowerCase());
  }

  if (prescription.frequency) {
    parts.push(prescription.frequency);
  }

  return parts.join(', ');
}

/**
 * Check if prescription is expired
 */
export function isExpired(prescription: Prescription): boolean {
  if (!prescription.end_date) return false;
  return new Date(prescription.end_date) < new Date();
}

/**
 * Check if prescription is expired (end date has passed)
 */
export function isExpiredPrescription(prescription: Prescription): boolean {
  if (!prescription.end_date) return false;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const endDate = new Date(prescription.end_date);
  endDate.setHours(0, 0, 0, 0);
  return endDate < today;
}

/**
 * Check if prescription is ending soon (within 7 days) but not yet expired
 */
export function isEndingSoon(prescription: Prescription): boolean {
  if (!prescription.end_date || prescription.status !== PrescriptionStatus.ACTIVE) {
    return false;
  }

  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const endDate = new Date(prescription.end_date);
  endDate.setHours(0, 0, 0, 0);

  // If already expired, not "ending soon"
  if (endDate < today) return false;

  const sevenDaysFromNow = new Date(today);
  sevenDaysFromNow.setDate(sevenDaysFromNow.getDate() + 7);

  return endDate <= sevenDaysFromNow;
}

/**
 * Get refill status for a prescription
 * Returns: 'can_refill' | 'needs_renewal' | 'expired' | null
 * - 'can_refill': Has refills remaining and ending/ended soon
 * - 'needs_renewal': No refills remaining and ending/ended soon (needs new prescription)
 * - 'expired': Prescription has passed its end date
 * - null: Not applicable (not ending soon, no end date, or not active)
 */
export function getRefillStatus(prescription: Prescription): 'can_refill' | 'needs_renewal' | 'expired' | null {
  if (prescription.status !== PrescriptionStatus.ACTIVE) {
    return null;
  }

  if (!prescription.end_date) {
    return null;
  }

  const isExpired = isExpiredPrescription(prescription);
  const endingSoon = isEndingSoon(prescription);

  if (isExpired) {
    // Prescription has expired
    if (prescription.refills > 0) {
      return 'can_refill'; // Can still refill even though expired
    }
    return 'expired'; // Expired and no refills - needs renewal
  }

  if (endingSoon) {
    // Ending within 7 days
    if (prescription.refills > 0) {
      return 'can_refill';
    }
    return 'needs_renewal';
  }

  return null;
}

/**
 * Check if prescription needs refill soon (within 7 days) - DEPRECATED
 * Use getRefillStatus() instead for more detailed information
 */
export function needsRefillSoon(prescription: Prescription): boolean {
  const status = getRefillStatus(prescription);
  return status === 'can_refill' || status === 'needs_renewal' || status === 'expired';
}

/**
 * Calculate remaining refills
 */
export function getRemainingRefills(prescription: Prescription): number {
  // This would need to track actual refill usage from backend
  // For now, just return the refills count
  return prescription.refills;
}

File: backend/src/models/prescription.rs
/*!
 * Prescription Model
 *
 * Represents a medication prescription for a patient.
 * All medication data fields are encrypted using AES-256-GCM before database storage.
 * Supports refill tracking, status management, and drug interaction warnings.
 */

use crate::utils::encryption::EncryptionKey;
use anyhow::{Context, Result};
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, PgPool, Type};
use uuid::Uuid;
use validator::Validate;

/// Prescription status enum
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Type)]
#[sqlx(type_name = "varchar", rename_all = "SCREAMING_SNAKE_CASE")]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum PrescriptionStatus {
    /// Prescription is active
    Active,
    /// Prescription has been completed (all refills used)
    Completed,
    /// Prescription was cancelled
    Cancelled,
    /// Prescription has been discontinued by provider
    Discontinued,
    /// Prescription is temporarily on hold
    OnHold,
}

impl PrescriptionStatus {
    /// Check if prescription can be refilled
    pub fn can_refill(&self) -> bool {
        matches!(self, PrescriptionStatus::Active)
    }

    /// Check if prescription can be discontinued
    pub fn can_discontinue(&self) -> bool {
        matches!(self, PrescriptionStatus::Active | PrescriptionStatus::OnHold)
    }
}

/// Medication form enum
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum MedicationForm {
    Tablet,
    Capsule,
    Liquid,
    Syrup,
    Suspension,
    Injection,
    Topical,
    Cream,
    Ointment,
    Gel,
    Patch,
    Inhaler,
    Drops,
    Suppository,
    Other,
}

/// Route of administration enum
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RouteOfAdministration {
    Oral,
    Topical,
    Intravenous,
    Intramuscular,
    Subcutaneous,
    Sublingual,
    Rectal,
    Inhalation,
    Ophthalmic,
    Otic,
    Nasal,
    Transdermal,
    Other,
}

/// Drug interaction warning structure (stored as JSONB)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DrugInteractionWarning {
    pub medication_name: String,
    pub severity: String, // "minor", "moderate", "major"
    pub description: String,
}

/// Prescription model - database representation with ENCRYPTED fields
/// Fields marked with 🔒 are encrypted in the database
#[derive(Debug, Clone, FromRow)]
pub struct Prescription {
    pub id: Uuid,

    // References
    pub visit_id: Option<Uuid>,
    pub visit_date: Option<NaiveDate>,
    pub patient_id: Uuid,
    pub provider_id: Uuid,

    // Medication Information (🔒 Encrypted)
    pub medication_name: String,  // 🔒 Encrypted
    pub generic_name: Option<String>, // 🔒 Encrypted
    pub dosage: String,           // 🔒 Encrypted
    pub form: Option<String>,
    pub route: Option<String>,

    // Dosing Instructions (🔒 Encrypted)
    pub frequency: String,        // 🔒 Encrypted
    pub duration: Option<String>, // 🔒 Encrypted
    pub quantity: Option<i32>,
    pub refills: i32,

    // Patient Instructions (🔒 Encrypted)
    pub instructions: Option<String>,    // 🔒 Encrypted
    pub pharmacy_notes: Option<String>,  // 🔒 Encrypted

    // Prescription Metadata
    pub prescribed_date: NaiveDate,
    pub start_date: Option<NaiveDate>,
    pub end_date: Option<NaiveDate>,

    // Status
    pub status: PrescriptionStatus,
    pub discontinuation_reason: Option<String>, // 🔒 Encrypted
    pub discontinued_at: Option<DateTime<Utc>>,
    pub discontinued_by: Option<Uuid>,

    // Refill Tracking
    pub refills_remaining: Option<i32>,
    pub last_refill_date: Option<NaiveDate>,

    // Drug Interaction Warnings (future enhancement)
    pub has_interactions: bool,
    pub interaction_warnings: Option<sqlx::types::JsonValue>,

    // E-Prescription (future enhancement)
    pub e_prescription_id: Option<String>,
    pub e_prescription_sent_at: Option<DateTime<Utc>>,
    pub e_prescription_status: Option<String>,

    // Audit fields
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
    pub updated_by: Option<Uuid>,
}

/// Prescription creation request (API input with decrypted data)
#[derive(Debug, Clone, Serialize, Deserialize, Validate)]
pub struct CreatePrescriptionRequest {
    pub visit_id: Option<Uuid>,

    #[validate(length(min = 36, max = 36, message = "patient_id must be a valid UUID"))]
    pub patient_id: String,

    #[validate(length(min = 36, max = 36, message = "provider_id must be a valid UUID"))]
    pub provider_id: String,

    // Medication Information
    #[validate(length(min = 1, max = 255, message = "Medication name must be 1-255 characters"))]
    pub medication_name: String,

    #[validate(length(max = 255, message = "Generic name too long (max 255 chars)"))]
    pub generic_name: Option<String>,

    #[validate(length(min = 1, max = 100, message = "Dosage must be 1-100 characters"))]
    pub dosage: String,

    pub form: Option<MedicationForm>,
    pub route: Option<RouteOfAdministration>,

    // Dosing Instructions
    #[validate(length(min = 1, max = 100, message = "Frequency must be 1-100 characters"))]
    pub frequency: String,

    #[validate(length(max = 100, message = "Duration too long (max 100 chars)"))]
    pub duration: Option<String>,

    #[validate(range(min = 1, max = 10000, message = "Quantity must be 1-10000"))]
    pub quantity: Option<i32>,

    #[validate(range(min = 0, max = 12, message = "Refills must be 0-12"))]
    pub refills: Option<i32>,

    // Patient Instructions
    #[validate(length(max = 5000, message = "Instructions too long (max 5000 chars)"))]
    pub instructions: Option<String>,

    #[validate(length(max = 5000, message = "Pharmacy notes too long (max 5000 chars)"))]
    pub pharmacy_notes: Option<String>,

    // Dates
    pub prescribed_date: Option<NaiveDate>,
    pub start_date: Option<NaiveDate>,
    pub end_date: Option<NaiveDate>,

    // Drug interaction warnings detected at creation time
    pub interaction_warnings: Option<Vec<DrugInteractionWarning>>,
}

/// Prescription update request (API input with decrypted data)
#[derive(Debug, Clone, Serialize, Deserialize, Validate)]
pub struct UpdatePrescriptionRequest {
    #[validate(length(max = 255, message = "Generic name too long (max 255 chars)"))]
    pub generic_name: Option<String>,

    #[validate(length(max = 100, message = "Dosage too long (max 100 chars)"))]
    pub dosage: Option<String>,

    pub form: Option<MedicationForm>,
    pub route: Option<RouteOfAdministration>,

    #[validate(length(max = 100, message = "Frequency too long (max 100 chars)"))]
    pub frequency: Option<String>,

    #[validate(length(max = 100, message = "Duration too long (max 100 chars)"))]
    pub duration: Option<String>,

    #[validate(range(min = 1, max = 10000, message = "Quantity must be 1-10000"))]
    pub quantity: Option<i32>,

    #[validate(range(min = 0, max = 12, message = "Refills must be 0-12"))]
    pub refills: Option<i32>,

    #[validate(length(max = 5000, message = "Instructions too long (max 5000 chars)"))]
    pub instructions: Option<String>,

    #[validate(length(max = 5000, message = "Pharmacy notes too long (max 5000 chars)"))]
    pub pharmacy_notes: Option<String>,

    pub status: Option<PrescriptionStatus>,

    #[validate(length(max = 1000, message = "Discontinuation reason too long (max 1000 chars)"))]
    pub discontinuation_reason: Option<String>,

    pub start_date: Option<NaiveDate>,
    pub end_date: Option<NaiveDate>,
}

/// Prescription response (API output with decrypted data)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PrescriptionResponse {
    pub id: Uuid,
    pub visit_id: Option<Uuid>,
    pub visit_date: Option<NaiveDate>,
    pub patient_id: Uuid,
    pub provider_id: Uuid,

    // Decrypted medication information
    pub medication_name: String,
    pub generic_name: Option<String>,
    pub dosage: String,
    pub form: Option<String>,
    pub route: Option<String>,

    // Decrypted dosing instructions
    pub frequency: String,
    pub duration: Option<String>,
    pub quantity: Option<i32>,
    pub refills: i32,

    // Decrypted patient instructions
    pub instructions: Option<String>,
    pub pharmacy_notes: Option<String>,

    // Dates
    pub prescribed_date: NaiveDate,
    pub start_date: Option<NaiveDate>,
    pub end_date: Option<NaiveDate>,

    // Status
    pub status: PrescriptionStatus,
    pub discontinuation_reason: Option<String>,
    pub discontinued_at: Option<DateTime<Utc>>,
    pub discontinued_by: Option<Uuid>,

    // Refill tracking
    pub refills_remaining: Option<i32>,
    pub last_refill_date: Option<NaiveDate>,

    // Warnings
    pub has_interactions: bool,
    pub interaction_warnings: Option<Vec<DrugInteractionWarning>>,

    // E-Prescription
    pub e_prescription_id: Option<String>,
    pub e_prescription_sent_at: Option<DateTime<Utc>>,
    pub e_prescription_status: Option<String>,

    // Audit
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
    pub updated_by: Option<Uuid>,
}

/// Medication search result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MedicationSearchResult {
    pub name: String,
    pub generic_name: Option<String>,
    pub common_dosages: Vec<String>,
    pub forms: Vec<String>,
}

impl Prescription {
    /// Decrypt all encrypted fields and convert to response
    pub fn decrypt(&self, encryption_key: &EncryptionKey) -> Result<PrescriptionResponse> {
        // Decrypt medication information
        let medication_name = encryption_key
            .decrypt(&self.medication_name)
            .context("Failed to decrypt medication_name")?;

        let generic_name = self
            .generic_name
            .as_ref()
            .map(|enc| encryption_key.decrypt(enc))
            .transpose()
            .context("Failed to decrypt generic_name")?;

        let dosage = encryption_key
            .decrypt(&self.dosage)
            .context("Failed to decrypt dosage")?;

        // Decrypt dosing instructions
        let frequency = encryption_key
            .decrypt(&self.frequency)
            .context("Failed to decrypt frequency")?;

        let duration = self
            .duration
            .as_ref()
            .map(|enc| encryption_key.decrypt(enc))
            .transpose()
            .context("Failed to decrypt duration")?;

        // Decrypt patient instructions
        let instructions = self
            .instructions
            .as_ref()
            .map(|enc| encryption_key.decrypt(enc))
            .transpose()
            .context("Failed to decrypt instructions")?;

        let pharmacy_notes = self
            .pharmacy_notes
            .as_ref()
            .map(|enc| encryption_key.decrypt(enc))
            .transpose()
            .context("Failed to decrypt pharmacy_notes")?;

        let discontinuation_reason = self
            .discontinuation_reason
            .as_ref()
            .map(|enc| encryption_key.decrypt(enc))
            .transpose()
            .context("Failed to decrypt discontinuation_reason")?;

        // Parse interaction warnings if present
        let interaction_warnings = if let Some(ref warnings) = self.interaction_warnings {
            Some(
                serde_json::from_value(warnings.clone())
                    .context("Failed to parse interaction warnings")?,
            )
        } else {
            None
        };

        Ok(PrescriptionResponse {
            id: self.id,
            visit_id: self.visit_id,
            visit_date: self.visit_date,
            patient_id: self.patient_id,
            provider_id: self.provider_id,
            medication_name,
            generic_name,
            dosage,
            form: self.form.clone(),
            route: self.route.clone(),
            frequency,
            duration,
            quantity: self.quantity,
            refills: self.refills,
            instructions,
            pharmacy_notes,
            prescribed_date: self.prescribed_date,
            start_date: self.start_date,
            end_date: self.end_date,
            status: self.status,
            discontinuation_reason,
            discontinued_at: self.discontinued_at,
            discontinued_by: self.discontinued_by,
            refills_remaining: self.refills_remaining,
            last_refill_date: self.last_refill_date,
            has_interactions: self.has_interactions,
            interaction_warnings,
            e_prescription_id: self.e_prescription_id.clone(),
            e_prescription_sent_at: self.e_prescription_sent_at,
            e_prescription_status: self.e_prescription_status.clone(),
            created_at: self.created_at,
            updated_at: self.updated_at,
            created_by: self.created_by,
            updated_by: self.updated_by,
        })
    }

    /// Check if prescription can be refilled
    pub fn can_refill(&self) -> bool {
        self.status.can_refill()
            && self.refills_remaining.map(|r| r > 0).unwrap_or(false)
    }

    /// Check if prescription can be discontinued
    pub fn can_discontinue(&self) -> bool {
        self.status.can_discontinue()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_prescription_status_can_refill() {
        assert!(PrescriptionStatus::Active.can_refill());
        assert!(!PrescriptionStatus::Completed.can_refill());
        assert!(!PrescriptionStatus::Cancelled.can_refill());
        assert!(!PrescriptionStatus::Discontinued.can_refill());
        assert!(!PrescriptionStatus::OnHold.can_refill());
    }

    #[test]
    fn test_prescription_status_can_discontinue() {
        assert!(PrescriptionStatus::Active.can_discontinue());
        assert!(!PrescriptionStatus::Completed.can_discontinue());
        assert!(!PrescriptionStatus::Cancelled.can_discontinue());
        assert!(!PrescriptionStatus::Discontinued.can_discontinue());
        assert!(PrescriptionStatus::OnHold.can_discontinue());
    }

    #[test]
    fn test_medication_form() {
        let form = MedicationForm::Tablet;
        assert_eq!(form, MedicationForm::Tablet);

        let form2 = MedicationForm::Injection;
        assert_eq!(form2, MedicationForm::Injection);
    }

    #[test]
    fn test_route_of_administration() {
        let route = RouteOfAdministration::Oral;
        assert_eq!(route, RouteOfAdministration::Oral);

        let route2 = RouteOfAdministration::Intravenous;
        assert_eq!(route2, RouteOfAdministration::Intravenous);
    }
}
